aioshumway/__init__.py

```diff
@@ -0,0 +1,165 @@
+
+    Args:
+        message (bytes): Message for FFWD agent.
+    """
+    def __init__(self, message):
+        self.message = message
+        self.transport = None
+
+    def connection_made(self, transport):
+        self.transport = transport
+        self.transport.sendto(self.message)
+        self.transport.close()
+
+    def connection_lost(self, exc):
+        pass
+
+    def error_received(self, exc):
+        logger.warning(f'Error when sending metric: {exc}\n'
+                       f'Metric value: {self.message}')
+
+
+class Meter(shumway.Meter):
+    """A single metric with updateable value."""
+    async def flush(self, coro):
+        """Pass map of data to coro for processing."""
+        await coro({
+            'key': self.key,
+            'attributes': self._attributes,
+            'value': self.value,
+            'type': 'metric',
+            'tags': self._tags
+        })
+
+
+class Counter(Meter, shumway.Counter):
+    """Keep track of an incrementing metric."""
+    pass
+
+
+class Timer(Meter, shumway.Timer):
+    """Time the execution of a block of code."""
+    async def __aenter__(self):
+        return self.__enter__()
+
+    async def __aexit__(self, exc_type, exc, tb):
+        self.__exit__(exc_type, exc, tb)
+
+
+class MetricRelay(shumway.MetricRelay):
+    """Manage creating and sending metrics to FFWD.
+
+    Args:
+        default_key (str): Key assigned to each metric.
+        ffwd_ip (str): IP address of FFWD agent.
+        ffwd_port (int): Port that FFWD agents listens on.
+        loop (obj): an asyncio.AbstractEventLoop-compatible event loop.
+    """
+
+    def __init__(self, default_key, ffwd_ip=FFWD_IP, ffwd_port=FFWD_PORT,
+                 loop=None):
+        super().__init__(default_key, ffwd_ip, ffwd_port)
+        if loop is None:
+            loop = asyncio.get_event_loop()
+        self.loop = loop
+
+    async def emit(self, metric, value, attributes=None, tags=None):
+        """Emit a single metric.
+
+        Args:
+            metric (str): Name of metric.
+            value (int/float): Value of metric.
+            attributes (dict): Additional attributes attached to metric.
+            tags (dict): Additional tags attached to metric.
+        """
+        one_time_metric = Meter(metric, key=self._default_key, value=value,
+                                attributes=attributes, tags=tags)
+        await self.flush_single(one_time_metric)
+
+    def incr(self, metric, value=1):
+        """Increment a metric. Create it if it does not exist.
+
+        Args:
+            metic (str): Name of metric.
+            value (int/float): Amount to increment the metric by.
+        """
+        try:
+            counter = self._metrics[metric]
+        except KeyError:
+            counter = Counter(metric, key=self._default_key)
+            self._metrics[metric] = counter
+        counter.incr(value)
+
+    def timer(self, metric):
```

Comment:
This and the `incr` methods didn't have to be overwritten, but it struck me that it was easy to make them more pythonic and follow the "easier to ask for forgiveness than permission" approach.